wallet/psbt.go

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2020 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package wallet
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcutil/psbt"
+	"github.com/btcsuite/btcwallet/waddrmgr"
+	"github.com/btcsuite/btcwallet/wallet/txauthor"
+	"github.com/btcsuite/btcwallet/wallet/txrules"
+	"github.com/btcsuite/btcwallet/wtxmgr"
+)
+
+// FundPsbt creates a fully populated PSBT packet that contains enough inputs to
+// fund the outputs specified in the passed in packet with the specified fee
+// rate. If there is change left, a change output from the wallet is added.
+//
+// NOTE: If the packet doesn't contain any inputs, coin selection is performed
+// automatically. If the packet does contain any inputs, it is assumed that full
+// coin selection happened externally and no additional inputs are added. If the
+// specified inputs aren't enough to fund the outputs with the given fee rate,
+// an error is returned. No lock lease is acquired for any of the
+// selected/validated inputs. It is in the caller's responsibility to lock the
+// inputs before handing them out.
+func (w *Wallet) FundPsbt(packet *psbt.Packet, account uint32,
+	feeSatPerKB btcutil.Amount) error {
+
+	// Make sure the packet is well formed. We only require there to be at
+	// least one output but not necessarily any inputs.
+	err := psbt.VerifyInputOutputLen(packet, false, true)
+	if err != nil {
+		return err
+	}
+
+	txOut := packet.UnsignedTx.TxOut
+	txIn := packet.UnsignedTx.TxIn
+
+	// Make sure none of the outputs are dust.
+	for _, output := range txOut {
+		// When checking an output for things like dusty-ness, we'll
+		// use the default mempool relay fee rather than the target
+		// effective fee rate to ensure accuracy. Otherwise, we may
+		// mistakenly mark small-ish, but not quite dust output as
+		// dust.
+		err := txrules.CheckOutput(output, txrules.DefaultRelayFeePerKb)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Let's find out the amount to fund first.
+	amt := int64(0)
+	for _, output := range txOut {
+		amt += output.Value
+	}
+
+	// addInputInfo is a helper function that fetches the UTXO information
+	// of an input and attaches it to the PSBT packet.
```

Comment:
So we assume all witness utxos? We _do_ currently support nested p2wkh which _technically_ is a non-witness utxo? 

Actually, I guess it isn't since there's no signature in the P2SH portion, and the witness signature will properly sign the input value.

Reply:
Nested P2WKH are witness outputs. The only difference on the signer side is that they have the `FinalScriptSig` set to a non-empty slice.

We will have to change this in a future bugfix to fix the PSBT security flaw: https://github.com/spesmilo/electrum/issues/6257#issuecomment-647736461 (also see https://github.com/lightningnetwork/lnd/issues/4400).

I'd rather not do it in this PR because a change in the `btcutil/psbt` library is also necessary to remove some of the sanity checks.

Reply:
Looking at `psbt/finalizer.go` a bit more, it looks like it expects you to only set the `RedeemScript`, then it'll construct the `sigScript` itself manually: https://github.com/btcsuite/btcutil/blob/master/psbt/finalizer.go#L417, then emplaced here: https://github.com/btcsuite/btcutil/blob/master/psbt/finalizer.go#L453